asllib/doc/SideEffects.tex

```diff
@@ -0,0 +1,698 @@
+\chapter{Side Effects\label{chap:SideEffects}}
+
+This chapter defines a static \emph{side effect analysis}.
+The analysis \emph{soundly} determines which pairs of expressions are \emph{order independent}.
+By \emph{soundly}, we mean that the analysis proves that a \underline{sufficient condition}
+for \emph{order independence} holds.
+\hypertarget{def-assertionerror}{}
+\[
+\assertionerror(C) \triangleq C \in \{ \DynamicErrorVal{\DynamicAssertionFailure}, \DynamicErrorVal{\DynamicTypeAssertionFailure} \}
+\]
+
+\begin{definition}[Order Independence]
+Two expressions $\veone$ and $\vetwo$ are \emph{order independent}
+if the following condition holds for every
+environment $\env \in \dynamicenvs$:
+\[
+\begin{array}{l}
+\{\ \abstractdynerror(C) \;|\; \evalexprlist{\env, [\veone, \vetwo]} \evalarrow C\ \} \setminus \{\bot\} = \\
+\{\ \abstractdynerror(C) \;|\; \reversevals(\evalexprlist{\env, [\vetwo, \veone]}) \evalarrow C\ \} \setminus \{\bot\} \enspace.
+\end{array}
+\]
+\hypertarget{def-reversevals}{}
+where $\reversevals$ reverses the order of values resulting from evaluating $\vetwo$ first
+and $\veone$ second into configurations resulting from evaluating $\veone$ first
+and $\vetwo$ second, leaving throwing configurations and dynamic errors unchanged:
+\[
+    \reversevals(C) \triangleq \begin{cases}
+        \Normal(([\vvone, \vvtwo], \vg), \env) & \text{if }C=\Normal(([\vvtwo, \vvone], \vg), \env)\\
+        C & \text{else}
+    \end{cases}
+\]
+\hypertarget{def-abstractdynerror}{}
+and $\abstractdynerror$ abstracts assertion errors into $\top$,
+all other dynamic errors into $\bot$ (which are ignored)
+and leaves all other configurations unchanged:
+\[
+\abstractdynerror(C) \triangleq \begin{cases}
+    C       & \text{if }\configdomain{C} \in \{\Normal, \Throwing\}\\
+    \top    & \text{if }\assertionerror(C)\\
+    \bot    & \text{else}\\
+\end{cases}
+\]
+That is, evaluating the expressions results with the same set of configurations,
+up to the order of values, and ignoring dynamic errors that are not due to assertions.\footnote{
+Ignoring dynamic errors that are not due to assertions is a limitation that will be lifted
+in a subsequent version of this reference.
+}
```

Comment:
please remove this footnote and either add it to ASLRefProgress if it is already in the code, or remove it entirely